aptos-move/framework/aptos-framework/sources/HashMap.move

```diff
@@ -0,0 +1,291 @@
+module AptosFramework::HashMap {
```

Comment:
Because this is based on Tables and is therefore large scale (expensive) storage, you should not call it Map (or HashMap).  The name Map is intended to be used for collection types which, similar like vectors, atomically fit into a resource.

Perhaps `BucketTable`?

Reply:
I have mixed feeling about this, developers need primitives that's scalable and cost efficient. Anything that fits into a single resource is not scalable. Table is not cost efficient since each entry occupies one storage slot. my opinion is we shouldn't expose Table to users, it's a low level memory allocator and we should build standard library based on it, but curious what's your thought

Reply:
We had many examples in DPN where maps/sets (regular ones, not storage based like table) sets would have been very helpful. We did have to use vectors for those use cases in DPN, which was ugly. Therefore the Move team is planning since a while to introduce those types, and they will be called `Map` and `Set`. Please do not collide with this naming. We have explicitly chosen the name `Table` for large storage collections to distinguish them from those planned map and set types.

Regards your concern about efficiency of `Table`, I'm not that sure about this. Many standard ERC contract use cases we have studied for EVM in the last few month use their `mapping` type, which is quite the same as our `Table`. They do not run in performance problems because one transaction typically works on only one (or maybe two) entries at a time. Your bucket table would be more inefficient for those cases, because you have to load the entire bucket to access a single entry, Plus the added logic for linear searching a bucket. 

I can imagine `BucketTable` (or however we call it) can be useful for some cases, but most use cases I have seen work quite well with `Table`.

Reply:
Another comment:

> developers need primitives that's scalable and cost efficient

A serious smart contract developer working with collections must be fully aware of whether they use large scale or small scale types. We can't abstract this away for them. I don't know how you calculate gas right now in Aptos, but a storage operation on EVM is 1000x more expensive than a plain memory operation. 

So the developer needs to be educated which type to use in which situation. The `BucketTable` may come in handy here for some use cases, but in others it would be way to inefficient/gas consuming,

Reply:
I know very little about smart contract development, this `BucketTable` is useful for implementing Staking module (https://github.com/aptos-labs/aptos-core/blob/dae89e8f650cfab0e7d97a5ce8c1357b309543f5/aptos-move/framework/aptos-framework/sources/configs/Stake.move#L58) for aptos. 

We need to support big number of delegations let's say 100k, while still want to iterate through all of them to distribute rewards with system transaction (in block metadata txn for example). if it's 100k storage slots, it'll be a massive writeset and waste of storage, if it's a single blob, the read/write amplification will be too big for any delegator.

Reply:
Sounds like a good application of the 'XYZTable'. 

Please note I like this new module. I'm just suggesting to not (a) name it '*Map' because we intent to use this name for something else in Move. Instead please use '*Table' (b) there is still need for tables like they are, and they are not just a lower level storage abstraction.

As regards iterability of tables, you need to very clearly tell developers that  this is a potentially dangerous feature. Maybe iterating you 100k still fits into one transaction, but my feeling is you are pretty close to the limit. So I'm not sure whether I think `Stake` is the right design for scalability.

Reply:
Agreed, I think we're on the same page, I'll update the PR with a different name (and other nits).

I'd like to discuss alternatives for `Stake` implementation sometime as well. Distributing rewards in a single transaction seems nice, for reference solana/avalanche does this with natives in rust/go. We can split it into multiple transactions as well but it feels less ergonomic and add more complexity.

Reply:
> Agreed, I think we're on the same page, I'll update the PR with a different name (and other nits).
> 
> I'd like to discuss alternatives for `Stake` implementation sometime as well. Distributing rewards in a single transaction seems nice, for reference solana/avalanche does this with natives in rust/go. We can split it into multiple transactions as well but it feels less ergonomic and add more complexity.

In general, I think things like Stake should go into an Aptos maintenance transaction which is scheduled repeatedly. This can be a general transaction `maintain_aptos` and it does some incremental work each time it is called, perhaps not only stake but other things. Just a thought; don't know much about the details here.